### PR TITLE
31 Ability to create lessons

### PR DESCRIPTION
**Pull request by ayomanetsan**
Implemented frontend and backend logic, so that a user can add lessons to the system. Due to the lack of design, I created a basic modal form component.

### CODE REVIEW

File: frontend/src/app/modules/lessons/lessons-create/lessons-create.component.ts [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        return this.myForm.get('date');
+    }
+
+    get questions() {
+        return this.myForm.get('questions');
+    }
+
+    get tags() {
+        return this.myForm.get('tags');
+    }
```
Comment (Anatolii-Ivanov) [inline frontend/src/app/modules/lessons/lessons-create/lessons-create.component.ts:83]: Do you need this?

File: frontend/src/app/services/lessons.service.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+@Injectable({
+    providedIn: 'root',
+})
+export class LessonsService {
+    public routePrefix = '/lessons';
+
+    constructor(private http: HttpService) { }
+
+    createLesson(lesson: INewLesson) {
+        this.http.post(this.routePrefix, lesson).subscribe();
```
Comment (Anatolii-Ivanov) [inline frontend/src/app/services/lessons.service.ts:14]: You should subscribe in component, not in a service

File: backend/EasySpeak.Core/EasySpeak.Core.WebAPI/Controllers/LessonsController.cs [hunk 3]
```diff
@@ -15,12 +16,31 @@ public LessonsController(ILessonsService lessonsService)
...
         {
             var lessons = await _lessonsService.GetAllLessonsAsync(filtersRequest);
 
             return Ok(lessons);
         }
+
+        [HttpPost]
+        public async Task<ActionResult<LessonDto>> CreateAsync(NewLessonDto lessonDto)
+        {
+            if (!ModelState.IsValid)
```
Comment (KaterynaHerasymiva) [inline backend/EasySpeak.Core/EasySpeak.Core.WebAPI/Controllers/LessonsController.cs:37]: As far as I know we don't need it https://learn.microsoft.com/en-us/aspnet/core/web-api/?view=aspnetcore-7.0#automatic-http-400-responses

File: frontend/src/app/modules/lessons/lessons-create/lessons-create.component.ts [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        'Design',
+        'History',
+        'Fashion',
+        'Medicine',
+        'Technologies',
+        'Pets',
+        'Philosophy',
+        'Photography',
+        'Politics',
+    ];
```
Comment (tatianahutii) [inline frontend/src/app/modules/lessons/lessons-create/lessons-create.component.ts:34]: Please move this  to .utils.ts

File: frontend/src/app/modules/lessons/lessons-create/lessons-create.component.ts [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+            date: ['', [
+                Validators.required,
+            ]],
+            questions: ['', [
+                Validators.required,
+            ]],
+            tags: ['', [
+                Validators.required,
+            ]],
+        });
```
Comment (tatianahutii) [inline frontend/src/app/modules/lessons/lessons-create/lessons-create.component.ts:62]: see the previous comments

File: frontend/src/app/modules/lessons/lessons-create/lessons-create.component.ts [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+    createLesson() {
+        const lessonQuestions: NewQuestion[] = [];
+
+        this.questions?.value
+            .split('\n')
+            .filter((entry: string) => entry.trim() !== '')
+            .forEach((element: string) => {
+                lessonQuestions.push(new NewQuestion(element, []));
+            });
```
Comment (tatianahutii) [inline frontend/src/app/modules/lessons/lessons-create/lessons-create.component.ts:93]: could you use map here?

File: frontend/src/app/modules/lessons/lessons-create/lessons-create.component.ts [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+
+        const lessonTags: NewTag[] = [];
+
+        this.tags?.value.forEach((element: string) => {
+            lessonTags.push(new NewTag(element));
+        });
```
Comment (tatianahutii) [inline frontend/src/app/modules/lessons/lessons-create/lessons-create.component.ts:99]: same here

File: frontend/src/app/modules/lessons/lessons.module.ts [hunk 8]
```diff
@@ -1,15 +1,34 @@
...
         CommonModule,
+        FormsModule,
+        ReactiveFormsModule,
+        MatButtonModule,
+        MatOptionModule,
+        MatDatepickerModule,
+        MatDialogModule,
+        MatFormFieldModule,
+        MatInputModule,
+        MatSelectModule,
```
Comment (tatianahutii) [inline frontend/src/app/modules/lessons/lessons.module.ts:31]: You need to import Material module

File: frontend/src/app/shared/models/lesson/NewQuestion.ts [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+
+export class NewQuestion {
+    topic: string;
+
+    subquestions: NewSubquestion[];
+
+    constructor(topic: string, subquestions: NewSubquestion[]) {
+        this.topic = topic;
+        this.subquestions = subquestions;
+    }
```
Comment (tatianahutii) [inline frontend/src/app/shared/models/lesson/NewQuestion.ts:11]: this is is not typical for ts

**Review by tatianahutii — CHANGES_REQUESTED** [review]

**Review by Anatolii-Ivanov — APPROVED** [review]

**Pull request merged**